File: KingSpade/KingSpades.py
debug = False
tests = [i + 1 for i in range(3)]
# problem may be able to be solved using variant of longest common substring
from pprint import pprint
import sys

faces = {"A", "K", "Q", "J", "T"}
suits = {"H", "D", "C", "S"}
d = {"total": 0, "A":0, "K":0, "Q":0, "J":0, "T":0, "H":0, "C":0, "D":0, "S":0, "KS":0}
for i in range(2,10):
    d[str(i)] = 0

def getPrefixDict(arr):
    d_prefix = [d.copy() for _ in range(len(arr) + 1)]

    # create and update 2 prefix arrays, one containing values, and the other containing dictionaries of seen cards
    for i in range(len(arr)):
        card = arr[i]
        # hard to filter over KS, but I could just ignore sets that have more than 0 KS.
        if card == "KS":
            d_prefix[i+1][card] += 1

        for k in d:
            d_prefix[i+1][k] += d_prefix[i][k]

        if card[0] in faces:
            if card[0] != "A":
                d_prefix[i+1]["total"] = d_prefix[i]["total"] + 10
            else:
                d_prefix[i+1]["total"] = d_prefix[i]["total"] + 1

        else:
            d_prefix[i+1]["total"] = d_prefix[i]["total"] + int(card[0])
        d_prefix[i+1][card[0]] += 1
        d_prefix[i+1][card[1]] += 1
    
    return d_prefix

# uses a sliding window to create a dictionary using values as keys, for each key contains an array of dictionaries where the card total is the value, and where there is not king of spades.
def getWindowDict(d_prefix, w_size):
    # perform a sliding window
    value_d = {}
    l = 0
    r = w_size
    while r < len(d_prefix):
        dict1 = d_prefix[r].copy()
        for k in dict1:
            dict1[k] -= d_prefix[l][k]
        if dict1["KS"] == 0:
            value_d[str(dict1.values())] = dict1["total"]

        l += 1
        r += 1
  
    return value_d

def findRes():
    n1, n2 = map(int, input().split())
    f_half = convert_to_lists(input().split())
    s_half = convert_to_lists(input().split())
    # print(f_half)
    # print(s_half)
    # print()

    res = 0
    f_value_d = {}
    s_value_d = {}
    for i in range(len(f_half)):
        f_li = f_half[i]
        f_prefix = getPrefixDict(f_li)
        w = min(100,len(f_li))
        while w > 0:
            f_value_d |= getWindowDict(f_prefix, w)
            w -= 1

    for j in range(len(s_half)):
        s_li = s_half[j]
        s_prefix =  getPrefixDict(s_li)
        w = min(100, len(s_li))
        
        while w > 0 and res // 2 <= w * 10:
            s_value_d |= getWindowDict(s_prefix, w)
            w -= 1
    for tup in f_value_d:
        if tup in s_value_d:
            res = max(res, f_value_d[tup]*2)
    print(res)

def test(test_nums):
    for t in test_nums:
        sys.stdin = open(f"input{t}.txt")
        findRes()
def convert_to_lists(li):
    new_list = []
    curr_li = []
    for i in range(len(li)):
        if li[i] == "KS" and len(curr_li) > 0:
            new_list.append(curr_li.copy())
            curr_li = []
        else:
            curr_li.append(li[i])
    if len(curr_li) > 0:
        new_list.append(curr_li.copy())
    return new_list

if debug:
    test(tests)
else:
    findRes()

File: KingSpade/KingSpades.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::collections::hash_map::HashMap;
use std::collections::hash_set::HashSet;

const DEBUG:bool = false;

struct Tools{
    face: HashSet<String>,
    d: HashMap<String, i32>,
    s_keys: Vec<String>
}

fn main() {
    let faces: HashSet<String> = HashSet::from(["A", "K", "Q", "J", "T"].map(|x| x.to_string()));
    let mut d: HashMap<String, i32> = HashMap::from([("total", 0), ("A",0), ("K",0), ("Q",0), ("J",0), ("T",0), ("H",0), ("C",0), ("D",0), ("S",0), ("KS",0)].map(|(x, y)| (x.to_string(), y)));
    let numbers: Vec<i32> = (2..10).collect();
    let mut s_keys:Vec<String> = d.keys().map(|x| x.to_string()).collect();
    s_keys.sort();
    for x in &numbers{
        d.insert(x.to_string(), 0);
    }
    let tools = Tools{face: faces.clone(),s_keys: s_keys.clone(), d: d.clone()};


    if DEBUG{
        test(&tools);
    }
    else{
        let lines = io::stdin();
        let mut line1 = String::new();
        let mut f_half_str = String::new();
        let mut s_half_str = String::new();
        let _ = lines.read_line(&mut line1);
        let _ = lines.read_line(&mut f_half_str);
        let _ = lines.read_line(&mut s_half_str);
        let _ = trim_newline(&mut f_half_str);
        let _ = f_half_str.trim();
        let _ = trim_newline(&mut s_half_str);
        let _ = s_half_str.trim();
        let f_half: Vec<String> = f_half_str.split(" ").map(|s| s.to_string()).collect();
        let s_half: Vec<String> = s_half_str.split(" ").map(|s| s.to_string()).collect();
        
        find_res(&tools, f_half, s_half);
    }
}

fn test(tools: &Tools){
    let tests: Vec<i32> = (1..6).collect();
    for t in tests{
        if let Ok(mut lines) = read_lines(format!("input{t}.txt")) {
            let _ = lines.next();

            let f_half_str = lines.next().unwrap().expect("First half of deck input string");
            let f_half: Vec<String> = f_half_str.split(" ").map(|s| s.to_string()).collect();

            let s_half_str = lines.next().unwrap().expect("s half");
            let s_half: Vec<String> = s_half_str.split(" ").map(|s| s.to_string()).collect();
            find_res(&tools, f_half, s_half);
            
        }
    } 
}
fn find_res(tools: &Tools, fh:Vec<String>, sh:Vec<String>){
    let f_prefix = get_prefix_dict(&tools, fh.clone());
    let s_prefix = get_prefix_dict(&tools, sh.clone());
    let mut res = 0;
    for w in 1..101{
        let f_value_d = get_window_dict(tools, f_prefix.clone(), w);
        let s_value_d = get_window_dict(tools, s_prefix.clone(), w);

        for s in f_value_d.keys(){
            if s_value_d.contains_key(s){
                res = std::cmp::max(res, f_value_d[s]*2);
            }
        }
    }

    println!("{}", res);
}

// The output is wrapped in a Result to allow matching on errors.
// Returns an Iterator to the Reader of the lines of the file.
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn get_prefix_dict(tools: &Tools, fh:Vec<String>) -> Vec<HashMap<String, i32>>{
    let mut d_prefix:Vec<HashMap<String, i32>> = vec![];
    for _ in 0..fh.len() + 1{
        d_prefix.push(tools.d.clone());
    }
    
    for i in 0..fh.len(){
        let card:String = fh[i].clone();
        let c1 = card.clone().chars().nth(0).unwrap().to_string();
        let c2 = card.clone().chars().nth(1).unwrap().to_string();
        if card == "KS".to_string(){
            *d_prefix[i+1].get_mut("KS").unwrap() = d_prefix[i+1].get("KS").unwrap() + 1;
        }

        for k in tools.d.keys(){
            *d_prefix[i+1].get_mut(k).unwrap() = d_prefix[i+1].get(k).unwrap() + d_prefix[i].get(k).unwrap();
        }

        if tools.face.contains(&c1){
            if c1 != "A".to_string(){
                *d_prefix[i+1].get_mut("total").unwrap() = d_prefix[i].get("total").unwrap() + 10;
            }
            else{
                *d_prefix[i+1].get_mut("total").unwrap() = d_prefix[i].get("total").unwrap() + 1;
            }
        }
        else{
            *d_prefix[i+1].get_mut("total").unwrap() = d_prefix[i].get("total").unwrap() + c1.parse::<i32>().unwrap();
        }
        *d_prefix[i+1].get_mut(&c1).unwrap() = d_prefix[i+1].get(&c1).unwrap() + 1;
        *d_prefix[i+1].get_mut(&c2).unwrap() = d_prefix[i+1].get(&c2).unwrap() + 1;
    }
    
    return d_prefix;
}

fn get_window_dict(tools: &Tools, d_prefix:Vec<HashMap<String, i32>>, w_size:i32)-> HashMap<String, i32>{
    let mut value_d: HashMap<String, i32> = HashMap::new();
    let mut l = 0;
    let mut r = w_size as usize;

    while r < d_prefix.len(){
        let mut dict1 = d_prefix[r].clone();
        for k in &tools.s_keys{
            *dict1.get_mut(k).unwrap() = dict1.get(k).unwrap() - d_prefix[l][k];
        }
        if dict1["KS"] == 0{
            let dict_str: &str = &tools.s_keys.iter()
                .map(|x| dict1[x].to_string())
                .collect::<Vec<_>>()
                .join(",");
            if !value_d.contains_key(dict_str){
                value_d.insert(dict_str.to_string(), dict1["total"]);
            }
        }
        l += 1;
        r += 1;
    }

    return value_d;
}
fn trim_newline(s: &mut String) {
    while s.ends_with('\n') || s.ends_with('\r') {
        s.pop();
    }
}

